### ISSUE TITLE
Ease debugging of "layout did not converge within 5 attempts" issues

### ISSUE THREAD

**Issue by oliverpool** [typst/typst#4528]
### Description

Add a way to reduce the number of layout attempts, to be able to output 5 pdfs (one for each iteration), allowing to better understand the lack of convergence.

When I fist encountered this error (see **Use Case**), my first thought was to augment the number of layout attempts (since it surely must converge at some point, right?), however after more thoughts, this would likely be bad idea, since people (like me) would just be lazy and increase this number without consideration.

Hence the opposite proposal: allow the number of attempts to be decreased.
This would allow me to see 5 PDFs and get a better feeling, why my layout is not converging.

### Use Case

I am trying to implement an `autobreak` feature (for a songbook system, see https://github.com/typst/typst/issues/993#issuecomment-2213532656), were I get `layout did not converge within 5 attempts` and no idea how to solve it:

https://typst.app/project/r9GASqmDOYasvWLDv5x4cI

If I execute the logic in my head, the layout should converge immediately, since at the beginning of every `autobreak` block, I can compute where the current block should go. So I really don't understand how Typst is computing the layout and why it is taking so many steps. Having a way to see the step 1-4 should help me get a better understanding (and eventually find a suitable fix).

---

BTW Typst looks awesome, thank your for this hard work!

I am creating a songbook-system to replace [patacrep](https://github.com/patacrep/patacrep) (which is based on LaTeX + [Songs](https://songs.sourceforge.net/)). The webapp looks really neat and I think that multiple people I know would love to use it to create their next songbook collaboratively (I know 3 songbook-projects which could benefit from it ;)

**Comment by laurmaedje** [comment]
Yeah, some way to debug this error would certainly be nice. It's a difficult problem, but at some time we'll need to figure something out.

**Comment by zinifier** [comment]
I had the same problem when trying to reach a specific number of pages. Usecase: i'm making booklets so i want to add pagebreaks so that the number of pages is a multiple of 4. It's unclear to me when the total page count is updated (on every layout iteration?) and how that affects dynamic page break insertion.

**Comment by oliverpool** [comment]
(Note: to not pollute this feature request, I opened a discussion dedicated to my specific problem and eventual solutions: #4530)

**Comment by Mc-Zen** [comment]
I suggest adding a CLI argument under the name `max-layout-iterations` or similar that restricts the number of layout iterations (maximum 5). This way a document can be compiled to see how it looks after a certain number of iterations. 

Note that this version is comparably straight-forward and has far less implementation overhead (and potential discussions) than the originally proposed solution of outputting 5 pdf for each iteration. Still, the same is basically achieved. 

For debugging purposes, it would of course be more direct to be able to change this setting right in the document, but I can also see some reasons against that.

**Comment by swaits** [comment]
> EDIT: This is not a blocker for me. I simply moved the page break logic to after typst compilation.

@zinifier how did you accomplish this? I'm having a similar problem now and seeking solutions.

**Comment by zinifier** [comment]
@swaits I add blank pages to the PDF after typst compilation. Not my favorite solution but it does the trick!

**Comment by swaits** [comment]
Not my favorite either, though I'm glad it helps you @zinifier.

Makes me wonder... why couldn't Typst just do the same thing, add extra pages, see if it converges, repeat a few times, then if converges just delete the extra pages before finalizing the PDF output?

It's all hacks, I agree... but this is a problem experienced by many Typstists. Any solution, temporary or otherwise, would be great.

Barring that, let us users decide on max iterations. I get that if it's not converging then some theoretical thing must be wrong. But that doesn't help me finish my already overdue papers. I, and others at my company, use Typst for real work now; and I'm quickly finding that I strongly depend on it.

@laurmaedje - how can I help here? I'm happy to legitimately make fixing this my top non-work tech priority.

**Comment by davystrong** [comment]
Was just about to create a feature request for this exact feature. I agree that allowing any value will just lead to me being lazy and increasing it to make it converge, which is a bad idea. But allowing me to set the number of iterations up to a maximum of 5 would REALLY help debugging, and seems like a fairly easy thing to implement (though I obviously don't know).

**Comment by laurmaedje** [comment]
Here are my initial collected thoughts on how we could improve the debugging situation. In part, this summarizes existing ideas, but there are also a few new ones.

### Locating the problem

Ideally, we'd have a span / error range for what specific introspection actually didn't converge. However, it's really quite difficult to implement with how memoization works. Still, somehow we need to solve this.

### Visualizing how the document evolves

It would be quite useful, as a user, to have a way to step through the different iterations to understand what's happening where (I think this should primarily be an IDE feature of the web app and tinymist). This should be fairly easy to implement and it's something that's on our web app roadmap.

### Analysing patterns in the state

We could try to analyse how the state evolves to give the user a bit more insight into what's going wrong.

I think the common cases of how the state evolves are

1. A, B, C, D, E, F, G, G (converges, but too late)
2. A, B, C, B, C, B, C, ... (doesn't coverge, but oscillates)
3. A, B, C, D, E, F, G, H, I, ... (doesn't converge, also doesn't oscillate, may still exhibit patterns)

To help the user, we could try to

- run a fixed number of additional iterations for diagnostic purposes (to understand whether it would converge but just not quickly enough), but still output the document from iteration 5. That would help us understand if the code does in principle converge, but too slowly (case 1). Might be frustrating though if Typst tells you "hey, it converged after 7 iterations, but I'm not gonna give you the final document".

- detect state oscillations to know for sure that it cannot converge (case 2).

- detect more patterns in the state (case 3) to predict whether it is likely to converge. Much fuzzier than the first two proposals. But if e.g. we have a state that evolves in a monotonic fashion, _perhaps_ it's possible to detect that.

**Comment by davystrong** [comment]
Having the compiler tell me where things went wrong would be great. However, given the complexity of most of these options, simply giving me the option to specify the max number of iterations (up to 5) seems like an important first step. I can then produce five PDFs and check where changes happen. It's not pretty but it seems like it should be simple to implement and it'll never have any uncommon cases where it won't work. It should also be useful long term since it would be easier to detect and report edge cases for better compiler hints.

**Comment by davystrong** [comment]
I created the [issue](https://github.com/Myriad-Dreamin/tinymist/issues/1307) in tinymist. I agree it would be perfect as an IDE feature but I'm not sure I agree it would be out of place in the CLI. Unlike autocompletion (which I know was an extreme example), this isn't something only related to writing the document. It is to do with compiling the document, and therefore including the flag in the compiler seems reasonable. I may be biased though because, while I use an IDE basically all the time, I don't like it when I HAVE to use the IDE. I like that all my tools are available as CLIs. So I'll not press it any further, it's up to you.

**Comment by davystrong** [comment]
@laurmaedje is this effectively the same thing (with the exception of the document now having an "__iterations" counter and the fact that the document is now inside a `context` environment)?
```typst
#let compiler(document, max-iterations: 5) = context {
  assert(max-iterations <= 5, message: "max-iterations must be <= 5")
  let iters = counter("__iterations")
  if iters.final().first() != 5 - max-iterations {
    iters.update(iters.final().first() + 1)
  } else {
    iters.update(5 - max-iterations)
    document
  }
}

#show: compiler.with(max-iterations: 2)
```

Or is there something I'm missing that also affects the number of iterations? This certainly acts as I'd expect it to when included at the start of the document, but I'm not sure if imports, for example, are imported regardless.

**Comment by laurmaedje** [comment]
This is cursed, I love it. I think it will probably do what you want. Imports will happen regardless, but that can't have any effect on the layout if you don't put anything into the document in your `show:` rule.

**Comment by davystrong** [comment]
I was considering creating a package but thought it might fall foul of the "Packages must not attempt to exploit the compiler" requirement. If not, I'll try to create a package at some stage when I have the time.

Also, credit where it's due, it was @Myriad-Dreamin who [suggested](https://github.com/Myriad-Dreamin/tinymist/issues/1307#issuecomment-2662690956) something like this might be possible.

**Comment by laurmaedje** [comment]
I wouldn't count that as an exploit. Exploit would be more in the direction of trying to break the sandbox or something like that.

**Comment by davystrong** [comment]
My [package](https://typst.app/universe/package/layout-ltd) to limit the number of compiler iterations is now available on typst Universe.

**Comment by swaits** [comment]
Nice! Next logical step.. script generating output with this package `for iterations in [1..=5]`, then create a progressive diff of the images from each iteration to the next.

**Comment by mattfbacon** [comment]
no comment on this issue but, @SolidTux `String::leak` is fine there because the strings are saved for the entire program runtime anyway so nothing is really leaked compared to `String`.

**Comment by laurmaedje** [comment]
Just opened https://github.com/typst/typst/pull/7364, which implements detailed non-convergence diagnostics.
